### PR TITLE
fix!: Use 32bit Vector2 in Flame to be compatible with shaders and forge2d

### LINKED ISSUE #3406: Stop using vector_math 64
### What could be improved

Stop using vector_math_64

### Why should this be improved

Because we don't need all that precision and it makes it annoying to work with numbers on low level features

### Risks

It may silently break existing games. 

### More information

_No response_

### Other

- [ ] Are you interested in working on a PR for this?

### PR DESCRIPTION
**Pull request by zeyus**
# Description

This fixes a problem where, if you use imports from both flame, and flame_forge2d that use Vector2 objects, the app will fail to compile (in addition to showing errors in the IDE) due to `flame` using `vector_math_64.dart` and `forge2d` / `flame_forge2d` using `vector_math.dart`). This occurs because of the merged PR: https://github.com/flame-engine/forge2d/pull/96

```dart
import 'package:flame/components.dart';
import 'package:flame_forge2d/flame_forge2d.dart';

class Ball extends BodyComponent<RiseTogetherGame> with ContactCallbacks {
  bool isMoving = false;
  bool isRising = false;
  final double radius;

  Ball({required this.radius, super.paint, Vector2? pos})
      : startPosition = pos != null ? Vector2.copy(pos) : Vector2.zero(), // <---- Error here
        super();
}
```

I would have kept this in a local branch but I saw at least one issue referencing this (including the Flame 2 issue): https://github.com/flame-engine/flame/issues/3406 and https://github.com/flame-engine/flame/issues/1938

Fixes #3406 

I'm not sure if this change breaks anything, I'm currently setting up / running the test suite, which so far is going fine, but there may be an issue with the PNG transform, perhaps someone else knows far better than I do.

I'll gladly add some tests for the pool, but before I want to do any work I would appreciate some feedback on:

a) if there is any chance this will be merged
b) if this is even the right approach

Thanks 🪄 

## Checklist

- [x] I have followed the [Contributor Guide] when preparing my PR.
- [ ] I have updated/added tests for ALL new/updated/fixed functionality.
- [ ] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- [ ] I have updated/added relevant examples in `examples` or `docs`.

## Breaking Change?

- [x] Yes, this PR is a breaking change. **Probably?**
- [ ] No, this PR is not a breaking change.

### Migration instructions

Change all manual imports that you have of `vector_math` from:
```
import 'package:vector_math/vector_math_64.dart';
```
to
```
import 'package:vector_math/vector_math.dart';
```

## Related Issues

[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Conventional Commit]: https://conventionalcommits.org
[CHANGELOG]: https://github.com/flame-engine/flame/blob/main/CHANGELOG.md

### CODE REVIEW

File: doc/tutorials/platformer/app/pubspec.yaml [hunk 1]
```diff
@@ -19,6 +19,5 @@ dev_dependencies:
     sdk: flutter
 
 flutter:
-  uses-material-design: true
```
Comment (spydon) [inline doc/tutorials/platformer/app/pubspec.yaml:22]: These should not be removed, the warnings you are seeing will be removed (and fixed) in the next Flutter release.

Comment (spydon) [reply doc/tutorials/platformer/app/pubspec.yaml:22]: This and it's sibling PRs is what we're waiting for:
https://github.com/flutter/flutter/pull/160443

Comment (zeyus) [reply doc/tutorials/platformer/app/pubspec.yaml:22]: Right, that makes more sense, I'll revert that commit.

The other test errors seem to be floating point math issues due to the change in precision. I didn't want to just change the tests to make them pass, but if is acceptable, maybe the tolerance can be increased a little

Comment (spydon) [reply doc/tutorials/platformer/app/pubspec.yaml:22]: Can you check how much the tolerance would have to change? It should just be the default of the closeToVector matcher (or whatever it is called).

Comment (zeyus) [reply doc/tutorials/platformer/app/pubspec.yaml:22]: Of course. I'll have a look when I get to work. I remember seeing something like 1e-10 (maybe it was more) in the test output, and the one fail that stood out to me was the PNG related one...I think I will try and manually inspect the output of that test to see what is going on.

Comment (zeyus) [reply doc/tutorials/platformer/app/pubspec.yaml:22]: > Can you check how much the tolerance would have to change? It should just be the default of the closeToVector matcher (or whatever it is called).

I'm going through it right now, and the current tolerance is set to 1e-14, and now it is within 1e-6 (actually ~9.5e-7) which looks to me to fit with reduction in precision. I'll adjust the relevant tolerances and see if anything still fails.

For completeness' sake, today I'm working on OSX and yesterday was Windows, and the results are pretty much the same.

Comment (zeyus) [reply doc/tutorials/platformer/app/pubspec.yaml]: > It should just be the default of the closeToVector matcher

So, the default `closeTo...` `delta` parameter is 1e-15 which is OK for 64 bit, but now it's less with 32 bit.

I've updated the relevant tests but there are some edge cases here where my intuition is reaching its limits and I don't know enough about how exactly dart handles things. That said, only 9 tests are failing now (huge improvement, still not great)

I'm fairly certain some of the problems are coming from toDouble in the Vector2 extension, because the doubles have 53 bits of precision, and conversion from a 32 bit float to a double will approximate a value...but that's kind of as far as I got with troubleshooting it...

Then there are 2 other "classes" of failures besides the toDouble failures:

- The image related issues (which are no doubt consequences of the change in precision)
- The "contains" fails, again, it's precision related, because if it checks for <= 32.0 and its 32.00001, then it fails

And a final weird one, that I don't even understand where it's coming from, because in `flame/test/components/position_component_test.dart:996:1` the failure shows the two values the same:

```text
00:17 +933 -13: /Users/au662726/github_projects/flame/packages/flame/test/components/position_component_test.dart: PositionComponent Bounding rectangle rotated component [E]                          
  Expected: Rect:<Rect.fromLTRB(-0.3, 0.0, 4.9, 2.8)>
    Actual: Rect:<Rect.fromLTRB(-0.3, 0.0, 4.9, 2.8)>
  
  package:matcher                                      expect
  test/components/position_component_test.dart 996:11  main.<fn>.<fn>.<fn>
```

Anyway that's where I'm at, and I am happy to dig further, but perhaps you have a suggestion as to how to deal with the float64 / double => float32 handling and rounding.

Comment (spydon) [reply doc/tutorials/platformer/app/pubspec.yaml]: > The image related issues (which are no doubt consequences of the change in precision) <-- I checked the output, and you can see that it's clearly 1px off, which is a rounding issue. Here's the isolatedDiff.png 

That is fine, we'll just regenerate the goldens.

> The "contains" fails, again, it's precision related, because if it checks for <= 32.0 and its 32.00001, then it fails

I guess we could add the epsilon here too.

> And a final weird one, that I don't even understand where it's coming from, because in flame/test/components/position_component_test.dart:996:1 the failure shows the two values the same:

That's a very odd one... No idea what is happening there.

Sorry for the late reply, I thought I already replied.

Comment (zeyus) [reply doc/tutorials/platformer/app/pubspec.yaml]: Great, I would love to help get this done (I'm currently using this branch in my experiment I'm developing). Please let me know what I can / should do to finish it up. 

> That's a very odd one... No idea what is happening there.

I figured it out, it's the same issue as the others, but the `toString()` method rounds the output: `pkg/sky_engine/lib/ui/geometry.dart`

```dart
class Rect {
  // ...
  /// The offset of the left edge of this rectangle from the x axis.
  final double left;

  /// The offset of the top edge of this rectangle from the y axis.
  final double top;

  /// The offset of the right edge of this rectangle from the x axis.
  final double right;

  /// The offset of the bottom edge of this rectangle from the y axis.
  final double bottom;

  // ...

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) {
      return true;
    }
    if (runtimeType != other.runtimeType) {
      return false;
    }
    return other is Rect &&
        other.left == left &&
        other.top == top &&
        other.right == right &&
        other.bottom == bottom;
  }

  @override
  int get hashCode => Object.hash(left, top, right, bottom);

  @override
  String toString() =>
      'Rect.fromLTRB(${left.toStringAsFixed(1)}, ${top.toStringAsFixed(1)}, ${right.toStringAsFixed(1)}, ${bottom.toStringAsFixed(1)})';
```

So despite it showing the same value in the output, the double values are not identical.

Comment (spydon) [reply doc/tutorials/platformer/app/pubspec.yaml]: I updated the goldens here: https://github.com/flame-engine/flame/pull/3515/commits/413282b934e9a051395db8711ca8467b0c0c18bb
To update them you just need to run `melos update-goldens`, the problem is that it needs to be done on a linux machine, since there are small discrepancies between the platforms in Flutter, and our pipeline is running Linux.

> I figured it out, it's the same issue as the others, but the toString() method rounds the output: pkg/sky_engine/lib/ui/geometry.dart

Aah, that makes sense!

Comment (spydon) [reply doc/tutorials/platformer/app/pubspec.yaml]: > Great, I would love to help get this done (I'm currently using this branch in my experiment I'm developing). Please let me know what I can / should do to finish it up.

Did you figure out what is left to do on this PR? :)

Comment (zeyus) [reply doc/tutorials/platformer/app/pubspec.yaml]: @spydon I'm not sure what's happening, but all the tests pass on my Linux laptop...I'm unsure why the tests would fail in Ubuntu on the github runner...

Comment (spydon) [reply doc/tutorials/platformer/app/pubspec.yaml]: The tests are failing locally on my linux box too, how are you running them?

Comment (zeyus) [reply doc/tutorials/platformer/app/pubspec.yaml]: @spydon i tried with `melos test` (and chose the flame) and same with `flutter test` within the  `packages/flame` directory.

also...to be sure, I did a `git diff` and there's nothing showing.

There's a potential difference here, because I'm heavily relying on `dart --enable-experiment=native-assets` i'm on the `main` channel of flutter/dart...I would be honestly surprised if that caused the difference though? 

I'll try change to the stable channel and report back

Comment (zeyus) [reply doc/tutorials/platformer/app/pubspec.yaml]: @spydon Ok, just confirmed, the tests pass on flutter/dart main channel, but not on stable (tested and failed on stable, changed back to main and they passed again).

That's interesting though...but at least I can try and get them passing on stable.

Comment (zeyus) [reply doc/tutorials/platformer/app/pubspec.yaml]: ...and now after switching back to stable, the error isn't showing up again...could it be something RNG related? 
I've tried `rm -rf packages/flame/.dart_tool` and re-running on flutter stable again, and it still passes..

nothing at all has changed on this machine since the last time I switched (not even a reboot...)

Comment (zeyus) [reply doc/tutorials/platformer/app/pubspec.yaml]: yeah confusing, just tried on my uni macbook and it fails (and I'm using flutter main on it)...obviously can't update the goldens but you already did that anyway, so I'll just use the macbook for now

Comment (zeyus) [reply doc/tutorials/platformer/app/pubspec.yaml]: @spydon I'm tracking this down, but I have a question.

In [`packages/flame/test/components/position_component_test.dart`](https://github.com/zeyus/flame/blob/feat/Vector2_32bit/packages/flame/test/components/position_component_test.dart#L214-L229), there are tests to make sure that a point on the edge is contained within the hitbox, e.g. "component with anchor bottomRight contains point on edge"

I need a bit of clarification as to how this works, because as far as I can tell, it actually shouldn't?

`expect(component.containsPoint(Vector2(1, 1)), true);` is the bottom right corner,

[packages/flame/lib/src/components/mixins/gesture_hitboxes.dart](https://github.com/zeyus/flame/blob/feat/Vector2_32bit/packages/flame/lib/src/components/mixins/gesture_hitboxes.dart#L18-L20)'s definition is:

```dart
bool containsPoint(Vector2 point) {
    return hitboxes.any((hitbox) => hitbox.containsPoint(point));
  }
```

then:

```dart
  bool containsPoint(Vector2 point) {
    return _possiblyContainsPoint(point) && super.containsPoint(point);
  }
```

LHS:
```dart
// vector_math Aabb2
  bool containsVector2(Vector2 other) =>
      (_min.x < other.x) &&
      (_min.y < other.y) &&
      (_max.x > other.x) &&
      (_max.y > other.y);
 ````
 RHS:
 ```dart
 // position_component.dart PositionComponent
   bool containsLocalPoint(Vector2 point) {
    return (point.x >= 0) &&
        (point.y >= 0) &&
        (point.x < _size.x) &&
        (point.y < _size.y);
  }
  ```
  
  So as far as I can tell, LHS will fail, because  `1.0 ≮ 1.0`, and if I'm understanding correctly then the RHS should fail as well (if it gets `x = 2.0, y=2.0, size=2.0; ∴ 2.0 ≮ 2.0` ).
 
 Am I understanding this correctly? Or am I barking up the wrong tree here? :)

Comment (spydon) [reply doc/tutorials/platformer/app/pubspec.yaml]: > Am I understanding this correctly? Or am I barking up the wrong tree here? :)

That does indeed seem like the case, good find! It might have been some float rounding errors before then...
Could you just change them to "close to the edge" and substract epsilon from the points?

Comment (zeyus) [reply doc/tutorials/platformer/app/pubspec.yaml]: ...which I could populate down the tree adding a method for intersectsLocalPoint, intersectsPoint, _possiblyIntersectsPoint, etc

Comment (spydon) [reply doc/tutorials/platformer/app/pubspec.yaml]: We could swap to using that for the `_possiblyContainsPoint`, but the RHS intentionally leaves the bottom and right as false on the `containsLocalPoint` so that if you have a grid it won't hit two cells when there is a point on the edge.

Comment (zeyus) [reply doc/tutorials/platformer/app/pubspec.yaml]: ok that makes sense, then if that's the expectation then I don't think it's logical to change `_possiblyContainsPoint`, I'll update the tests instead

Comment (zeyus) [reply doc/tutorials/platformer/app/pubspec.yaml]: Ok, seems to be heading in the right direction. I'll continue tomorrow, gotta catch some sleep!

Comment (spydon) [reply doc/tutorials/platformer/app/pubspec.yaml]: > [could not help myself...is that it?](https://github.com/flame-engine/flame/pull/3515/commits/84883a8bcf6a38bb7e709b90ffc003d30ce2341c)

Haha, very good job! I'll review it tomorrow, I too need to catch some sleep.

File: packages/flame/test/game/game_render_box_test.dart [hunk 2]
```diff
@@ -9,6 +9,8 @@ class _MockFlameGame extends Mock implements FlameGame {}
 
 class _MockBuildContext extends Mock implements BuildContext {}
 
+// Ignoring dart linting error due
+// ignore: subtype_of_base_or_final_is_not_base_final_or_sealed,invalid_use_of_type_outside_library
```
Comment (spydon) [inline packages/flame/test/game/game_render_box_test.dart:13]: This was on the main channel right? I think we can remove it until it is released.

File: packages/flame/test/math/vector2_32bit_test.dart.txt [hunk 3]
```diff
@@ -0,0 +1 @@
+import 'package:flame/math.dart';
```
Comment (spydon) [inline packages/flame/test/math/vector2_32bit_test.dart.txt:1]: Empty file

File: packages/flame_test/lib/src/epsilon.dart [hunk 4]
```diff
@@ -0,0 +1,110 @@
+import 'dart:typed_data';
+
+/// Returns the next larger representable float32 value from [value].
+///
+/// This is useful for creating test values that are just above a target value.
+/// For example, if you want a value that's just above 1.0, you can use:
+/// `final justAbove1 = getNextFloat32(1.0);`
+double getNextFloat32(double value) {
```
Comment (spydon) [inline packages/flame_test/lib/src/epsilon.dart:8]: Generally in Dart you don't name things `get`/`set`, this could just be `nextFloat32`. :)

Very nice helper methods though!

File: packages/flame_test/lib/src/epsilon.dart [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+  // Use bit-level inspection
+  final data = ByteData(8);
+  data.setFloat64(0, value);
+  // The sign bit is the most significant bit
+  return (data.getUint8(0) & 0x80) != 0;
+}
+
+/// The machine epsilon for float32 (IEEE 754 single precision)
+/// This is the difference between 1.0 and the next representable float32 value
+const double float32Epsilon = 1.1920928955078125e-7; // 2^-23
```
Comment (spydon) [inline packages/flame_test/lib/src/epsilon.dart:106]: I guess these could be used in all the tests that manually specify an epsilon now?

**Review by spydon — COMMENTED** [review]
Great job! Just some minor comments

File: packages/flame/test/math/vector2_32bit_test.dart.txt (hunk 3 above)
Comment (zeyus) [reply packages/flame/test/math/vector2_32bit_test.dart.txt:1]: haha what in the world...this must have been some 1am trash...I can't even remember the last time I made a txt file

**Comment by zeyus** [comment]
OK so that was:
- A lot more changes than I was expecting; and
- Uncovered another potential issue in: `test/collisions/collision_type_test.dart 193:9`
    I'm not sure exactly what the root cause is...I'm running the tests again now (which intermittently pass/fail...curse you RNG gods), but I think it's the size of the generated blocks so I'm trying with them set to `prevFloat32(10)`

File: packages/flame/test/components/position_component_test.dart [hunk 6]
```diff
@@ -221,15 +229,35 @@ void main() {
...
 
-          expect(component.containsPoint(Vector2(1, 1)), true);
-          expect(component.containsPoint(Vector2(1, -1)), true);
-          expect(component.containsPoint(Vector2(-1, -1)), true);
-          expect(component.containsPoint(Vector2(-1, 1)), true);
+          expect(
+            component.containsPoint(
+              Vector2(prevFloat32(1.0), prevFloat32(1.0)),
+            ),
+            true,
```
Comment (spydon) [inline packages/flame/test/components/position_component_test.dart:236]: Use the `isTrue` matcher for all `true` checks in `expect`.

File: packages/flame_test/test/epsilon_test.dart [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+    test('NaN stays NaN', () {
+      const nan = double.nan;
+      expect(nextFloat32(nan).isNaN, isTrue);
+      expect(prevFloat32(nan).isNaN, isTrue);
+    });
+
+    test('infinite stays inf, next(-inf) and prev(inf) are finite', () {
+      const posInf = double.infinity;
+      const negInf = double.negativeInfinity;
```
Comment (spydon) [inline packages/flame_test/test/epsilon_test.dart:153]: We're trying to stay away from abbreviations in the code-base, even though these are quite easy to understand what they mean.

**Comment by zeyus** [comment]
OK fixed review changes, and indeed, the prevFloat32(10) for the sizing issue fixed the test.

I also had RNG problems with the position_component_test...I'm not sure if my calculations on the cumulative tolerance is exactly correct, but I think it's right? obviously the tolerance depends on the numbers within the vectors, but from when it was failing they were ~3e-5

File: packages/flame/test/components/position_component_test.dart [hunk 8]
```diff
@@ -221,15 +229,35 @@ void main() {
...
+            component.containsPoint(
+              Vector2(prevFloat32(1.0), prevFloat32(1.0)),
+            ),
+            isTrue,
+          );
+          expect(
+            component.containsPoint(
+              Vector2(prevFloat32(1.0), nextFloat32(-1.0)),
+            ),
+            true,
```
Comment (spydon) [inline packages/flame/test/components/position_component_test.dart:242]:
```suggestion
            isTrue,
```

File: packages/flame/test/components/position_component_test.dart [hunk 9]
```diff
@@ -221,15 +229,35 @@ void main() {
...
+          );
+          expect(
+            component.containsPoint(
+              Vector2(nextFloat32(-1.0), nextFloat32(-1.0)),
+            ),
+            true,
```
Comment (spydon) [inline packages/flame/test/components/position_component_test.dart:248]:
```suggestion
            isTrue,
```

File: packages/flame/test/components/position_component_test.dart [hunk 10]
```diff
@@ -221,15 +229,35 @@ void main() {
...
+          );
+          expect(
+            component.containsPoint(
+              Vector2(nextFloat32(-1.0), prevFloat32(1.0)),
+            ),
+            true,
```
Comment (spydon) [inline packages/flame/test/components/position_component_test.dart:254]:
```suggestion
            isTrue,
```

File: packages/flame/test/components/position_component_test.dart [hunk 11]
```diff
@@ -238,10 +266,33 @@ void main() {
...
-          expect(component.containsPoint(Vector2(-1, -1)), true);
-          expect(component.containsPoint(Vector2(-1, 1)), true);
+          expect(
+            component.containsPoint(
+              Vector2(
+                prevFloat32(1.0),
+                prevFloat32(1.0),
+              ),
+            ),
+            true,
```
Comment (spydon) [inline packages/flame/test/components/position_component_test.dart:276]:
```suggestion
            isTrue,
```

File: packages/flame/test/components/position_component_test.dart [hunk 12]
```diff
@@ -238,10 +266,33 @@ void main() {
...
+          );
+          expect(
+            component.containsPoint(
+              Vector2(prevFloat32(1.0), nextFloat32(-1.0)),
+            ),
+            true,
```
Comment (spydon) [inline packages/flame/test/components/position_component_test.dart:282]:
```suggestion
            isTrue,
```

File: packages/flame/test/components/position_component_test.dart [hunk 13]
```diff
@@ -238,10 +266,33 @@ void main() {
...
+          );
+          expect(
+            component.containsPoint(
+              Vector2(nextFloat32(-1.0), nextFloat32(-1.0)),
+            ),
+            true,
```
Comment (spydon) [inline packages/flame/test/components/position_component_test.dart:288]:
```suggestion
            isTrue,
```

File: packages/flame/test/components/position_component_test.dart [hunk 14]
```diff
@@ -238,10 +266,33 @@ void main() {
...
+          );
+          expect(
+            component.containsPoint(
+              Vector2(nextFloat32(-1.0), prevFloat32(1.0)),
+            ),
+            true,
```
Comment (spydon) [inline packages/flame/test/components/position_component_test.dart:294]:
```suggestion
            isTrue,
```

File: packages/flame/test/components/position_component_test.dart [hunk 15]
```diff
@@ -646,7 +745,7 @@ void main() {
           ..position = Vector2(50, 20)
           ..anchor = const Anchor(0.1, 0.2);
         parent.add(child);
-
+        // var tolerance = toleranceVector2Float32(child.position);
```
Comment (spydon) [inline packages/flame/test/components/position_component_test.dart:748]: Not needed?

File: packages/flame/test/game/transform2d_test.dart [hunk 16]
```diff
@@ -170,19 +171,50 @@ void main() {
...
+            point2.y,
+            (toleranceFloat32(point1.y) +
+                    toleranceFloat32(point2.y) * 2 +
+                    toleranceVector2Float32(translation) +
+                    toleranceFloat32(rotation) +
+                    toleranceVector2Float32(scale) +
+                    toleranceVector2Float32(offset)) *
+                4,
+          ),
+        );
```
Comment (spydon) [inline packages/flame/test/game/transform2d_test.dart:217]: Would it be easier to use `closeToVector` here?

Comment (zeyus) [reply packages/flame/test/game/transform2d_test.dart:217]: yeah, of course :) I think I broke it out when didn't have the toleranceVector2... helper.

**Review by renancaraujo — COMMENTED** [review]
We tried this change in the past. The reasons I decided to not go trough with are related to big imprecisoons generated by sequential changes on transformation matrices. 

Detailed comment is here: https://github.com/flame-engine/flame/pull/3405#issuecomment-2545600637

**Comment by spydon** [comment]
@zeyus are you on our [Discord](https://discord.com/invite/pxrBmy4)? :)

**Comment by zeyus** [comment]
@spydon I am now!

**Review by spydon — APPROVED** [review]
I have tried this out on my game today and I haven't noticed any problems at all, just a few exports that had to change (added that to the migration instructions).
Great job @zeyus!
